DECOMPRESSION FPGA DESIGN (GZIP and SNAPPY) (#7)
tyoungsc/oneAPI-samples: DirectProgramming/DPC++FPGA/ReferenceDesigns/gunzip/README.md

> # GZIP Decompression

Reviewer: General comment on this file: there was a recent PR for all the FPGA Readmes: https://github.com/oneapi-src/oneAPI-samples/pull/858  Can you make sure these changes are incorporated into this README?

Author: I checked out the recent change. Does all of that information really belong in every README? It is generic to all programs, could it not go in a higher level README?

> # GZIP Decompression

Reviewer: Please have someone in Tech Comms review this README

> This DPC++ reference design demonstrates an GZIP decompression design on an FPGA.

Reviewer: "This gzip decompression design does gzip decompression" :-)  How about this:
```suggestion
This DPC++ reference design implements GZIP decompression on an FPGA.  This includes LZ77 decoding and Huffman decoding with static or dynamic huffman tables. 
```

* [DPC++ FPGA Code Samples Guide](https://software.intel.com/content/www/us/en/develop/articles/explore-dpcpp-through-intel-fpga-code-samples.html) helps you to navigate the samples and build your knowledge of DPC++ for FPGA. <br>
* [oneAPI DPC++ FPGA Optimization Guide](https://software.intel.com/content/www/us/en/develop/documentation/oneapi-fpga-optimization-guide) is the reference manual for targeting FPGAs through DPC++. <br>
* [oneAPI Programming Guide](https://software.intel.com/en-us/oneapi-programming-guide) is a general resource for target-independent DPC++ programming. 
 
| Optimized for                     | Description
---                                 |---
| OS                                | Ubuntu* 18.04/20.04, RHEL*/CentOS* 8, SUSE* 15; Windows* 10
| Hardware                          | Intel&reg; Programmable Acceleration Card (PAC) with Intel Arria&reg; 10 GX FPGA <br> Intel&reg; FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix&reg; 10 SX) <br> Intel Xeon&reg; CPU E5-1650 v2 @ 3.50GHz (host machine)
| Software                          | Intel&reg; oneAPI DPC++ Compiler <br> Intel&reg; FPGA Add-On for oneAPI Base Toolkit

> | What you will learn               | Creating an efficient GZIP decompression engine on an FPGA.

Reviewer: ```suggestion
| What you will learn               | How to implement an efficient GZIP decompression engine on an FPGA.
```

| Time to complete                  | 1 hour

## Purpose

> This FPGA reference design demonstrates an efficient GZIP decompression engine. See the [Additional Design Information Section](#additional-design-information) for more information on GZIP and DEFLATE de/compression for the FPGA.

Reviewer: ```suggestion
This FPGA reference design demonstrates an efficient GZIP decompression engine. See the [Additional Design Information Section](#additional-design-information) for more information on GZIP and DEFLATE compression and decompression.
```

     ```

### Example of Output
You should see output similar to the following in the console:
```
>>>>> Uncompressed File Test <<<<<
Decompressing '../data/uncompressed.gz' 1 time
Launching kernels for run 0
All kernels have finished for run 0
>>>>> Uncompressed File Test: PASSED <<<<<

>>>>> Statically Compressed File Test <<<<<
Decompressing '../data/static_compressed.gz' 1 time
Launching kernels for run 0
All kernels have finished for run 0
>>>>> Statically Compressed File Test: PASSED <<<<<

>>>>> Dynamically Compressed File Test <<<<<
Decompressing '../data/dynamic_compressed.gz' 1 time
Launching kernels for run 0
All kernels have finished for run 0
>>>>> Dynamically Compressed File Test: PASSED <<<<<

>>>>> Throughput Test <<<<<
Decompressing '../data/tp_test.gz' 5 times
Launching kernels for run 0
All kernels have finished for run 0
Launching kernels for run 1
All kernels have finished for run 1
Launching kernels for run 2
All kernels have finished for run 2
Launching kernels for run 3
All kernels have finished for run 3
Launching kernels for run 4
All kernels have finished for run 4
Execution time: 7.12956 ms
Output Throughput: 561.045 MB/s

> Compression Ratio: 1021.26

Reviewer: Maybe this should be expressed explicitly as a ratio (xxx:1)?

Also, is this number right?  That file is compressed 1021:1 ???

Author: I can switch to a percentage, or use a ratio. What do you think?
and yes. It is a max throughput test.

>>>>> Throughput Test: PASSED <<<<<

PASSED
```

NOTE: When running on the FPGA emulator, the *Execution time* and *Throughput* do not reflect the design's actual hardware performance. <br>

> NOTE: The throughput of the decompression engine is highly dependant on the compression ratio of the input file, more specifically on the amount of LZ77 compression. The throughput test reports the *maximum* throughput, since it is decompressing a highly compressed file. When decompressing other files, it is expected that the throughput will be lower.

Reviewer: ```suggestion
NOTE: The throughput of the decompression engine is highly dependant on the compression ratio of the input file, more specifically on the amount of LZ77 compression. The file chosen for this test has a very high amount of LZ77 compression, and was chosen to demonstrate the *maximum* throughput achievable with this implementation. When decompressing other files, the throughput may be somewhat lower.
```
Can we put a lower bound on how much lower?  'In the worst case, as much as xx% (50%?) lower'

Author: The lowest would be a completely uncompressed file, which would be like 90% decrease.

## Additional Design Information
### Source Code Breakdown
The following source files can be found in the `src/` sub-directory.

| File                            | Description 
|:---                             |:---

> |`main.cpp`                       | Contains the `main()` function and the top-level launching, validation, and performance measurements.

Reviewer: ```suggestion
|`main.cpp`                       | Contains the `main()` function which launches the kernels, validates the results, and measures performance.
```

> |`byte_bit_stream.hpp`            | A bitstream class that has 8-bits (a byte) written in and a dynamic number of bits read.

Reviewer: ```suggestion
|`byte_bit_stream.hpp`            | A bitstream class that accepts one byte (8 bits) at a time and allows a variable number of bits to be read out on each transaction.
```

> |`byte_stacker.hpp`               | A kernel that accepts between 0 and N elements per cycle, buffers N elements locally, and writes out N elements when possible.

Reviewer: ```suggestion
|`byte_stacker.hpp`               | A kernel that accepts between 0 and N elements per cycle and combines them to output N elements at a time.
```

> |`common.hpp`                     | Functions common across the design.

Reviewer: '...such as. . .'

|`gzip_decompressor.hpp`          | The top-level file for the GZIP decompressor. This file launches all of the GZIP kernels.
|`gzip_header_data.hpp`           | A class to store the GZIP header data.
|`gzip_metadata_reader.hpp`       | A kernel that streams in a GZIP file, parses and strips the GZIP header and footer metadata, and streams the payload into the DEFLATE decompressor engine.
|`huffman_decoder.hpp`            | A kernel that implements Huffman decoding. It streams in DEFLATE blocks, a byte at a time, and streams out either a literal (character) or a {length, distance} pair.
|`lz77_decoder.hpp`               | A kernel that implements LZ77 decoding. It streams in a union of a literal (character) or a {length, distance} pair and streams out literals.
|`simple_crc32.hpp`               | A simple implementation of CRC-32 calculation. This is used to validate the output of the GZIP engine.

### GZIP and DEFLATE

> GZIP is specific implementation of the DEFLATE protocol. The structure of a GZIP file is illustrated in the figure below. It starts with a variable length, byte-aligned header of 10 or more bytes. Then follows the data payload, which is 1 or more DEFLATE compressed blocks. After the DEFLATE blocks, there may be some small amount of padding to realign to a byte boundary, and finally an 8-byte GZIP footer, which contains the CRC-32 and size (in bytes) of the uncompressed data. For more details on the GZIP file format, [see here](https://en.wikipedia.org/wiki/Gzip)

Reviewer: ```suggestion
GZIP is a specific implementation of the DEFLATE protocol. The structure of a GZIP file is illustrated in the figure below. It starts with a variable length, byte-aligned header of 10 or more bytes. Then follows the data payload, which is 1 or more DEFLATE compressed blocks. After the DEFLATE blocks, there may be some small amount of padding to realign to a byte boundary, and finally an 8-byte GZIP footer, which contains the CRC-32 and size (in bytes) of the uncompressed data. For more details on the GZIP file format, [see here](https://en.wikipedia.org/wiki/Gzip)
```

<img src="gzip_file_format.png" alt="gzip_file_format" width="400"/>

> The DEFLATE compression algorithm performs LZ77 Encoding followed by Huffman encoding (decompression decodes in the opposite order). For more information on the DEFLATE format and LZ77 and Huffman encoding/decoding, [start here](https://en.wikipedia.org/wiki/Deflate).

Reviewer: ```suggestion
The DEFLATE compression algorithm performs LZ77 Encoding followed by Huffman encoding. Therefore, decompression decodes in the opposite order. For more information on the DEFLATE format and LZ77 and Huffman encoding/decoding, [start here](https://en.wikipedia.org/wiki/Deflate).
```

  - **Statically compressed blocks** have a *static* Huffman table; a table defined in the DEFLATE format. Therefore the remainder of the data in the block is the payload to be decoded using this table.
  - **Dynamically compressed blocks** are illustrated in the figure below. The second table is a list of code lengths for the literal and distance symbols. It is used to create the Huffman table for decoding the payload data. The first table is used to encode the second table. So, to decode a dynamically compressed block, the reader must read the first table, then decode the second table using the first table, create the Huffman table from the code lengths of the second table, and finally decode the payload using the Huffman table.

The GZIP decompression engine in this reference design supports all three block types.

<img src="deflate_dynamic_block_format.png" alt="deflate_dynamic_block_format" width="600"/>

### GZIP Decompression FPGA Design

> The image below summarizes the full streaming GZIP decompression design. The orange kernels on the FPGA in the dashed box are the *actual* streaming GZIP decompression engine. The *Producer* and *Consumer* kernels are only used to stream data into and out of the GZIP decompression engine from device memory, respectively.

Reviewer: ```suggestion
The image below summarizes the full streaming GZIP decompression design. The orange kernels on the FPGA in the dashed box make up the streaming GZIP decompression engine. The *Producer* and *Consumer* kernels are only used to stream data into and out of the GZIP decompression engine from and to device memory, respectively.
```

<img src="gzip_decompression_engine_full.png" alt="gzip_decompression_engine_full" width="700"/>

> The figure below illustrates the GZIP decompression engine in isolation. It delineates the streaming GZIP decompressor from the streaming DEFLATE decompressor. The `GZIPMetaDataReader` kernel parses and strips away the GZIP header and footer data and forwards the consecutive DEFLATE blocks to the DEFLATE decompression engine.

Reviewer: I think I made this comment on your slides as well, but I'd just add the DEFLATE dotted line box to the diagram above and skip this second diagram.  Your call though.

<img src="gzip_decompression_engine_ip.png" alt="gzip_decompression_engine_ip" width="700"/>

The following subsections will briefly discuss the high-level details of each of the kernels in the GZIP decompression engine.

#### GZIP Metadata Reader
The GZIP metadata reader streams in in the GZIP file, a byte at a time, parses and strips away the GZIP header and footer, and forwards the remaining data to the DEFLATE portion of the decompression engine. The output of the GZIP Metadata Reader is a stream of DEFLATE format compressed blocks.

#### Huffman Decoder
The Huffman decoder streams in DEFLATE compressed blocks, a byte at a time, and streams out either literals (8-bit characters) or {length, distance} pairs. The Huffman decoder is not byte-aligned; it decodes the input stream bits at a time. To do this efficiently, the decoder uses the `ByteBitStream` class, which streams in bytes at a time but gives the decoder access to all of the bits in parallel. This design takes advantage of the fact that the DEFLATE format limits code lengths to at most 15 bits.

> The structure of the Huffman decoder is shown in the image below. The decoder examines the next 15 bits in the bit stream and checks for matches of length 1-15 in parallel to create a 15-bit *valid bitmap*, where a 1 at bit `i` indicates that the code of length `i` matches a code in the Huffman table. For example, a value of `0b000000000000101` would indicate that there is a match of length 1 and 3.

Reviewer: ```suggestion
The structure of the Huffman decoder is shown in the image below. The decoder examines the next 15 bits in the bit stream and checks for matches in the Huffman table of lengths 1-15 in parallel to create a 15-bit *valid bitmap*, where a 1 at bit `i` indicates that the code of length `i` matches a code in the Huffman table. For example, a value of `0b000000000000101` would indicate that there is a match of length 1 and 3.
```

> To decode the next symbol, the decoder finds the *shortest* of the 15 matches that were computed in parallel. To do so, it performs a `CTZ` (**c**ount **t**railing **z**eros) operation on the valid bitmap and adds one to it, which is equivalent to finding the shortest match length. Once the shortest match length is found, the bit stream is shifted by that many bits, the symbol for that match is looked up, and the process repeats. If the decoded symbol is a literal, then the literal is written to the output. If the decoded symbol is a length, then no output is generated in this iteration; the length is saved and the next iteration of the loop decodes the distance and generates the {length, distance} pair output.

Reviewer: ```suggestion
To decode the next symbol, the decoder finds the *shortest* of the 15 matches that were computed in parallel. To do so, it performs a `CTZ` (**c**ount **t**railing **z**eros) operation on the valid bitmap and adds one to it, which is equivalent to finding the shortest match length. Once the shortest match length is found, the bit stream is shifted by that many bits, the symbol for that match is determined from the Huffman table, and the process repeats. If the decoded symbol is a literal, then the literal is written to the output. If the decoded symbol is a length, then no output is generated in this iteration; the length is saved and the next iteration of the loop decodes the distance and generates the {length, distance} pair output.
```

```
match = (C >= first_code[L]) && (C < last_code[L]);
offset = C - first_code[L];
symbol = lit_map[base_idx[L] + offset]
```

The Huffman decoder decodes a literal in 1 iteration of the main loop, and a {length, distance} pair in 2 iterations. When decoding a {length, distance} pair, the length code can be 1-15 bits and, based on the decoded length, 0-5 extra bits. Similarly, the distance code can be 1-15 bits and 0-15 extra bits. To decode a {length, distance} in 2 iterations, the Huffman decoder looks at 30 bits in parallel. The first 15 are examined as described earlier. The decoder also examines the 15x5 and 15x15 different possibilities for the extra length and distance bits, respectively. For the extra length bits, the shortest matching code can be 1-15 bits, giving 15 possibilities for where the extra bits could *start*. There can be 1-5 extra bits, giving 5 possibilities, which yields 15x5 possibilities for the extra length bits. Decoding the extra distance bits uses the same approach but with 0-15 extra bits, yielding 15x15 possibilities.

#### LZ77 Decoder

> The input to the LZ77 decoder is a stream of commands that are either an 8-bit literal or a {length, distance} pair; the output of the Huffman decoder. The output is a stream of literals (8-bit characters). The LZ77 decoder keeps a 32K *history buffer*; a history of the last 32K literals it has streamed to the output. 

Reviewer: Is 'commands' the right term here?  Maybe 'symbols' is better?
```suggestion
The input to the LZ77 decoder is a stream of symbols(?) that are either an 8-bit literal or a {length, distance} pair. The output is a stream of literals (8-bit characters). The LZ77 decoder keeps a 32KB *history buffer* which stores the last 32K literals it has streamed to the output. 
```

<img src="lz77_decoder.png" alt="lz77_decoder" width="700"/>

When reading out of the history buffers, the *Current History Buffer Index* is again used to shuffle the output of the `N` buffers so that they are streamed out in the correct order. The following example illustrates reading the history buffer for `N=4`.

Let's assume that to start the LZ77 decoder received 14 consecutive literals and wrote them to the output and history buffers in cyclical order (i.e. to buffers `0, 1, 2, 3, 0, 1, 2, 3, ...`).
```

> Characters:    E T H O M E P H O N E

Reviewer: Really?  ET HOME PHONE, not ET PHONE HOME?

Author: https://www.youtube.com/watch?v=iyFijjikkeM

```
input = {'M', 'E', 'H', 'O'}
shuffle_vector = {2, 3, 0, 1}
output = shuffle(input, shuffle_vector)
       = shuffle({'M', 'E', 'H', 'O'}, {2, 3, 0, 1})
       = {'H', 'O', 'M', 'E'}
```

Now, the state looks like this:
```
Characters:    E T H O M E P H O N E H O M E
               _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
Buffer index:  0 1 2 3 0 1 2 3 0 1 2 3 0 1 2 0
```

In general, this process is repeated `ceil(length/N)` times. When `length` is not a multiple of `N`, the `valid_count` is used to tell the downstream kernel how many of the `N` literals are valid. There is an edge case when the `distance` in the {length, distance} pair is less than `N`. This is handled in the code, but is omitted in this document for brevity.

> The history buffers have a read-after-write dependency since, when reading from the history buffers, the LZ77 decoder is also writing back to them. The history buffers are implemented in on-chip RAMs and therefore have a read latency that results in an increased loop II. To fix this II issue, we use a shift-register cache to track in-flight writes to the history buffers. This technique is described in more detail in the **Caching On-Chip Memory to Improve Loop Performance** (onchip_memory_cache) tutorial.

Reviewer: Please include a link to the tutorial.

Author: We haven't done in any other reference designs that cite other tutorials. E.g. https://github.com/oneapi-src/oneAPI-samples/blob/master/DirectProgramming/DPC%2B%2BFPGA/ReferenceDesigns/qrd/README.md

Should these all be changed?

Reviewer: Your call now that you own the samples.  Whatever you decide, I made a similar comment in @yuguen-intel 's review on his fork, so let him know either way
